File: benchmarks/weird_captcha_gym/shared_runtime/app/mechanics/dead_mans_switch.js
(() => {
  "use strict";

  const KEY_TO_DIRECTION = {
    w: "N", arrowup: "N",
    d: "E", arrowright: "E",
    s: "S", arrowdown: "S",
    a: "W", arrowleft: "W",
  };
  const DELTAS = {N: [0, -1], E: [1, 0], S: [0, 1], W: [-1, 0]};
  let model = null;
  let activeCleanup = null;

  function clonePoint(point) {
    return {x: Number(point.x), y: Number(point.y)};
  }

  function samePoint(a, b) {
    return Number(a.x) === Number(b.x) && Number(a.y) === Number(b.y);
  }

  function eventTime() {
    return Math.round(performance.now() - model.startedAt);
  }

  function pushEvent(record) {
    const event = {seq: model.events.length + 1, t_ms: eventTime(), ...record};
    model.events.push(event);
    return event;
  }

  function pointKey(point) {
    return `${Number(point.x)},${Number(point.y)}`;
  }

  function pressureCenter(tMs = eventTime()) {
    const motion = model.state.pressure_motion;
    const phase = Number(motion.phase_milliradians) / 1000;
    const angle = Number(tMs) / Number(motion.period_ms) * Math.PI * 2 + phase;
    return {
      x: .5 + Number(motion.x_amplitude_milli) / 1000 * Math.sin(angle),
      y: .5 + Number(motion.y_amplitude_milli) / 1000 * Math.sin(angle * 2 + phase * .63),
    };
  }

  function normalizedPointer(event) {
    const track = document.querySelector(".pressure-track");
    const rect = track?.getBoundingClientRect();
    if (!rect) return null;
    return {
      x: Math.max(0, Math.min(1, (event.clientX - rect.left) / rect.width)),
      y: Math.max(0, Math.min(1, (event.clientY - rect.top) / rect.height)),
    };
  }

  function pressureContains(pointer, tMs = eventTime()) {
    if (!pointer) return false;
    const center = pressureCenter(tMs);
    const motion = model.state.pressure_motion;
    const dx = (Number(pointer.x) - center.x) / (Number(motion.hit_x_milli) / 1000);
    const dy = (Number(pointer.y) - center.y) / (Number(motion.hit_y_milli) / 1000);
    return dx * dx + dy * dy <= 1;
  }

  function updatePressureMotion() {
    if (!model) return;
    const tMs = eventTime();
    const center = pressureCenter(tMs);
    const pad = document.querySelector(".pressure-pad");
    if (pad) {
      pad.style.left = `${center.x * 100}%`;
      pad.style.top = `${center.y * 100}%`;
    }
    if (model.holding && !model.completed) {
      const inside = pressureContains(model.lastPointer, tMs);
      if (inside) {
        model.outsideSince = null;
        if (tMs - model.lastPressureSample >= Number(model.state.pressure_motion.sample_ms)) {
          pushEvent({type: "hold_sample", pointer: {x: Math.round(model.lastPointer.x * 1000), y: Math.round(model.lastPointer.y * 1000)}});
          model.lastPressureSample = tMs;
          model.pressureSamples += 1;
        }
      } else {
        if (model.outsideSince == null) model.outsideSince = tMs;
        if (tMs - model.outsideSince > Number(model.state.pressure_motion.outside_grace_ms)) {
          resetSuccessfulRun("pointerleave");
        }
      }
      maybeComplete();
    }
    const continuity = document.querySelector(".pressure-continuity i");
    const continuityText = document.querySelector(".pressure-continuity b");
    if (continuity && continuityText) {
      const started = model.events.find((event) => event.seq === model.currentHoldStartSeq);
      const elapsed = model.holding && started ? Math.max(0, tMs - Number(started.t_ms)) : 0;
      continuity.style.width = `${Math.min(100, elapsed / Number(model.state.pressure_motion.minimum_hold_ms) * 100)}%`;
      continuityText.textContent = `${(elapsed / 1000).toFixed(1)} / ${(Number(model.state.pressure_motion.minimum_hold_ms) / 1000).toFixed(1)}s`;
    }
    model.motionRaf = requestAnimationFrame(updatePressureMotion);
  }

  function installDeveloperReveal(state) {
    const form = document.getElementById("cheat-form");
    const input = document.getElementById("cheat-password");
    const output = document.getElementById("cheat-output");
    if (!form || !input || !output) return;
    form.addEventListener("submit", async (event) => {
      event.preventDefault();
      output.textContent = "";
      try {
        const response = await fetch("/cheat", {
          method: "POST",
          headers: {"content-type": "application/json"},
          body: JSON.stringify({password: input.value}),
        });
        if (!response.ok) {
          output.textContent = response.status === 404 ? "Disabled." : "Denied.";
          return;
        }
        const data = await response.json();
        output.textContent = `Route: ${(data.solution_path || []).join(" ")} · ${Number(data.minimum_success_moves || 0)} accepted moves`;
      } catch (_error) {
        output.textContent = "Unavailable.";
      }
    });
  }

  function updateScene() {
    if (!model) return;
    const {board} = model.state;
    const root = document.querySelector(".dead-switch-captcha");
    const vehicle = document.querySelector(".switch-vehicle");
    if (!root || !vehicle) return;
    root.dataset.holding = String(model.holding);
    root.dataset.completed = String(model.completed);
    vehicle.style.left = `${((model.position.x + 0.5) / Number(board.columns)) * 100}%`;
    vehicle.style.top = `${((model.position.y + 0.5) / Number(board.rows)) * 100}%`;
    vehicle.dataset.heading = model.heading;
    document.querySelectorAll(".switch-checkpoint").forEach((node) => {
      const order = Number(node.dataset.order || 0);
      node.dataset.passed = String(order <= model.checkpointIndex);
    });
    const pressureState = document.querySelector(".pressure-state");
    if (pressureState) pressureState.textContent = model.holding ? "CIRCUIT CLOSED" : "CIRCUIT OPEN";
    const progress = document.querySelector(".switch-route-progress b");
    if (progress) progress.textContent = `${model.checkpointIndex} / ${board.checkpoints.length}`;
    const resets = document.querySelector(".switch-reset-count b");
    if (resets) resets.textContent = String(model.resetCount).padStart(2, "0");
  }

  function showTerminalVerdict(kind, title, detail) {
    const root = document.querySelector(".dead-switch-captcha");
    const verdict = root?.querySelector(".switch-terminal-verdict");
    if (!root || !verdict) return;
    root.classList.toggle("is-passed", kind === "pass");
    root.classList.toggle("is-failed", kind === "fail");
    verdict.innerHTML = `<b>${title}</b><span>${detail}</span>`;
    if (kind === "fail") {
      window.setTimeout(() => root.classList.remove("is-failed"), 1350);
    }
  }

  function clearTerminalVerdict() {
    const root = document.querySelector(".dead-switch-captcha");
    root?.classList.remove("is-failed");
    const verdict = root?.querySelector(".switch-terminal-verdict");
    if (verdict) verdict.replaceChildren();
  }

  function resetSuccessfulRun(reason) {
    if (!model || !model.holding || model.completed) return;
    pushEvent({type: "hold_end", reason, position: clonePoint(model.position)});
    model.holding = false;
    model.pointerId = null;
    model.position = clonePoint(model.state.board.start);
    model.checkpointIndex = 0;
    model.currentHoldStartSeq = null;
    model.lastPointer = null;
    model.lastPressureSample = 0;
    model.outsideSince = null;
    model.resetCount += 1;
    model.helpers.setReadout(`RESET ${String(model.resetCount).padStart(2, "0")} · HOLD AGAIN`, "error");
    const course = document.querySelector(".switch-course");
    if (course) {
      course.classList.remove("is-resetting");
      void course.offsetWidth;
      course.classList.add("is-resetting");
    }
    updateScene();
  }

  async function submitCompletion() {
    if (!model || model.submitting || model.completed) return;
    model.submitting = true;
    model.completed = true;
    const submittedAt = eventTime();
    const holdStartEvent = model.events.find((event) => event.seq === model.currentHoldStartSeq);
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      completed: true,
      holding: true,
      events: model.events,
      path: model.events.filter((event) => event.type === "move").map((event) => event.direction),
      visited_checkpoints: model.state.board.checkpoints.slice(0, model.checkpointIndex).map((item) => item.id),
      reset_count: model.resetCount,
      final_position: clonePoint(model.position),
      successful_hold_start_seq: model.currentHoldStartSeq,
      submitted_t_ms: submittedAt,
      continuous_hold_duration_ms: holdStartEvent ? submittedAt - Number(holdStartEvent.t_ms) : 0,
      pressure_sample_count: model.pressureSamples,
    };
    updateScene();
    model.helpers.setReadout("DOCK HANDSHAKE…", "idle");
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify(payload),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        showTerminalVerdict("pass", "PASS", "CONTINUITY CONFIRMED");
        model.helpers.setReadout("PASS · CONTINUITY CONFIRMED", "passed");
      } else if (outcome.passed === false) {
        if (outcome.state) await render(outcome.state, model.helpers);
        showTerminalVerdict("fail", "FAIL", "COURSE REISSUED");
        model.helpers.setReadout("FAIL · COURSE REISSUED", "error");
      }
    } catch (_error) {
      model.completed = false;
      model.submitting = false;
      model.helpers.setReadout("LINK LOST · HOLD AND RETRY", "error");
      updateScene();
    }
  }

  function maybeComplete() {
    if (!model || !model.holding || model.completed || model.submitting) return;
    const board = model.state.board;
    if (!samePoint(model.position, board.goal) || model.checkpointIndex !== board.checkpoints.length) return;
    const holdStartEvent = model.events.find((event) => event.seq === model.currentHoldStartSeq);
    const duration = holdStartEvent ? eventTime() - Number(holdStartEvent.t_ms) : 0;
    if (duration < Number(model.state.pressure_motion.minimum_hold_ms)) {
      model.helpers.setReadout("DOCK READY · KEEP TRACKING UNTIL CONTINUITY FILLS", "pending");
      return;
    }
    if (eventTime() - model.lastPressureSample > Number(model.state.pressure_motion.maximum_sample_gap_ms)) return;
    submitCompletion();
  }

  function move(direction) {
    if (!model || !model.holding || model.completed) return;
    const before = clonePoint(model.position);
    const [dx, dy] = DELTAS[direction];
    const candidate = {x: before.x + dx, y: before.y + dy};
    const {board} = model.state;
    const accepted = (
      candidate.x >= 0 && candidate.x < Number(board.columns)
      && candidate.y >= 0 && candidate.y < Number(board.rows)
      && !model.walls.has(pointKey(candidate))
    );
    if (accepted) {
      model.position = candidate;
      model.heading = direction;
      const expected = board.checkpoints[model.checkpointIndex];
      if (expected && samePoint(candidate, expected)) {
        model.checkpointIndex += 1;
        model.helpers.setReadout(`CHECKPOINT ${model.checkpointIndex} LATCHED`, "idle");
      } else {
        model.helpers.setReadout("PRESSURE STABLE", "idle");
      }
    } else {
      model.helpers.setReadout("BULKHEAD · ROUTE BLOCKED", "error");
    }
    pushEvent({
      type: "move",
      direction,
      from: before,
      to: clonePoint(model.position),
      accepted,
      checkpoint_index: model.checkpointIndex,
    });
    updateScene();
    if (accepted && samePoint(model.position, board.goal) && model.checkpointIndex === board.checkpoints.length) {
      maybeComplete();
    }
  }

  async function abandonCourse() {
    if (!model || model.submitting || model.completed) return;
    model.submitting = true;
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify({mechanic_id: model.state.mechanic_id, task_id: model.state.task_id, challenge_id: model.state.challenge_id, completed: false, events: []}),
      });
      const outcome = await response.json();
      if (outcome.state) await render(outcome.state, model.helpers);
      showTerminalVerdict("fail", "FAIL", "NEW COURSE ISSUED");
      model.helpers.setReadout("FAIL · NEW COURSE ISSUED", "error");
    } catch (_error) {
      model.submitting = false;
      model.helpers.setReadout("REISSUE FAILED", "error");
    }
  }

  async function render(state, helpers) {
    if (activeCleanup) activeCleanup();
    document.body.dataset.mechanic = "dead-mans-switch";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    const walls = new Set((state.board.walls || []).map(pointKey));
    model = {
      state,
      helpers,
      startedAt: performance.now(),
      events: [],
      position: clonePoint(state.board.start),
      checkpointIndex: 0,
      resetCount: 0,
      holding: false,
      completed: false,
      submitting: false,
      pointerId: null,
      currentHoldStartSeq: null,
      lastPointer: null,
      lastPressureSample: 0,
      outsideSince: null,
      pressureSamples: 0,
      motionRaf: 0,
      heading: "E",
      walls,
    };

    const cellRecords = [];
    for (let y = 0; y < Number(state.board.rows); y += 1) {
      for (let x = 0; x < Number(state.board.columns); x += 1) {
        const wall = walls.has(`${x},${y}`);
        const checkpoint = state.board.checkpoints.find((item) => Number(item.x) === x && Number(item.y) === y);
        const start = Number(state.board.start.x) === x && Number(state.board.start.y) === y;
        const goal = Number(state.board.goal.x) === x && Number(state.board.goal.y) === y;
        cellRecords.push(`<div class="switch-cell${wall ? " is-wall" : ""}${checkpoint ? " switch-checkpoint" : ""}${start ? " is-start" : ""}${goal ? " is-goal" : ""}" data-order="${checkpoint ? checkpoint.order : ""}" data-passed="false">${checkpoint ? `<span>${checkpoint.order}</span>` : ""}${goal ? "<i>DOCK</i>" : ""}</div>`);
      }
    }

    helpers.app.innerHTML = `
      <section class="dead-switch-captcha" data-holding="false" data-completed="false">
        <div class="switch-terminal-verdict" aria-live="assertive"></div>
        <header class="dead-switch-head">
          <div><span>CONTINUITY TRIAL / HOLD-07</span><h1>${helpers.text(state.prompt)}</h1></div>
          <div class="switch-live-lamp"><i></i><b>DEAD-MAN LINE</b></div>
        </header>
        <section class="dead-switch-workbench">
          <div class="switch-course-wrap">
            <div class="switch-course" style="--switch-columns:${Number(state.board.columns)};--switch-rows:${Number(state.board.rows)}">
              <div class="switch-course-grid">${cellRecords.join("")}</div>
              <div class="switch-vehicle" data-heading="E"><i></i><b></b><span></span></div>
            </div>
            <div class="switch-route-legend"><span>ORDERED RELAYS</span><b>${state.board.checkpoints.map((item) => item.order).join(" → ")} → DOCK</b></div>
          </div>
          <aside class="pressure-console">
            <div class="pressure-gauge"><i></i><b></b><span class="pressure-state">CIRCUIT OPEN</span></div>
            <div class="pressure-track"><button type="button" class="pressure-pad" aria-label="Track and hold the moving pressure plate"><span>KEEP<br>DEPRESSED</span><i></i><b></b></button><div class="pressure-continuity"><span></span><i></i><b>0.0 / ${(Number(state.pressure_motion.minimum_hold_ms) / 1000).toFixed(1)}s</b></div></div>
            <p>Follow the moving plate while held.<br>Steer with WASD or arrows.</p>
            <div class="switch-console-stats"><span class="switch-route-progress">RELAYS <b>0 / ${state.board.checkpoints.length}</b></span><span class="switch-reset-count">RESETS <b>00</b></span></div>
          </aside>
        </section>
        <footer class="dead-switch-foot"><div class="readout" data-status="idle">PRESS AND HOLD TO ARM</div><button type="button" class="switch-abandon">REISSUE COURSE</button></footer>
        ${helpers.cheatPanelTemplate()}
      </section>
    `;

    const pad = document.querySelector(".pressure-pad");
    const keydown = (event) => {
      const direction = KEY_TO_DIRECTION[String(event.key || "").toLowerCase()];
      if (!direction || event.repeat || !model || model.completed) return;
      event.preventDefault();
      if (!model.holding) {
        helpers.setReadout("PRESSURE REQUIRED", "error");
        return;
      }
      move(direction);
    };
    const pointerdown = (event) => {
      if (!model || model.completed || model.holding || event.button !== 0) return;
      event.preventDefault();
      clearTerminalVerdict();
      model.holding = true;
      model.pointerId = event.pointerId;
      model.lastPointer = normalizedPointer(event);
      if (!pressureContains(model.lastPointer)) { model.holding = false; model.pointerId = null; return; }
      try { pad.setPointerCapture(event.pointerId); } catch (_error) { /* capture is best effort */ }
      const started = pushEvent({type: "hold_start", position: clonePoint(model.position), pointer: {x: Math.round(model.lastPointer.x * 1000), y: Math.round(model.lastPointer.y * 1000)}});
      model.currentHoldStartSeq = started.seq;
      model.lastPressureSample = started.t_ms;
      model.outsideSince = null;
      helpers.setReadout("CIRCUIT CLOSED · STEER NOW", "idle");
      updateScene();
    };
    const pointermove = (event) => {
      if (!model?.holding || event.pointerId !== model.pointerId || model.completed) return;
      model.lastPointer = normalizedPointer(event);
    };
    const pointerup = (event) => {
      if (model?.holding && event.pointerId === model.pointerId) resetSuccessfulRun("pointerup");
    };
    const pointercancel = (event) => {
      if (model?.holding && event.pointerId === model.pointerId) resetSuccessfulRun("pointercancel");
    };
    const lostcapture = () => {
      if (model?.holding && !model.completed) resetSuccessfulRun("lostcapture");
    };
    const blur = () => {
      if (model?.holding && !model.completed) resetSuccessfulRun("blur");
    };
    pad.addEventListener("pointerdown", pointerdown);
    pad.addEventListener("lostpointercapture", lostcapture);
    window.addEventListener("pointermove", pointermove);
    window.addEventListener("pointerup", pointerup);
    window.addEventListener("pointercancel", pointercancel);
    window.addEventListener("keydown", keydown);
    window.addEventListener("blur", blur);
    document.querySelector(".switch-abandon")?.addEventListener("click", abandonCourse);
    installDeveloperReveal(state);
    activeCleanup = () => {
      if (model?.motionRaf) cancelAnimationFrame(model.motionRaf);
      pad.removeEventListener("pointerdown", pointerdown);
      pad.removeEventListener("lostpointercapture", lostcapture);
      window.removeEventListener("pointermove", pointermove);
      window.removeEventListener("pointerup", pointerup);
      window.removeEventListener("pointercancel", pointercancel);
      window.removeEventListener("keydown", keydown);
      window.removeEventListener("blur", blur);
    };
    updateScene();
    model.motionRaf = requestAnimationFrame(updatePressureMotion);
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.dead_mans_switch = {rootSelector: ".dead-switch-captcha", render};
})();

File: benchmarks/weird_captcha_gym/shared_runtime/server/incubator_graders/dead_mans_switch.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "dead_mans_switch"
_MOVES = {
    "N": (0, -1),
    "E": (1, 0),
    "S": (0, 1),
    "W": (-1, 0),
}
_RELEASE_REASONS = {"pointerup", "pointercancel", "pointerleave", "lostcapture", "blur"}


def _failure(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": 0, "feedback": message}


def _point(value: Any) -> tuple[int, int] | None:
    if not isinstance(value, dict):
        return None
    try:
        return int(value["x"]), int(value["y"])
    except (KeyError, TypeError, ValueError):
        return None


def _time(value: Any) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(number) or number < 0 or number > 3_600_000:
        return None
    return number


def _integer(value: Any) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _normalized_pointer(value: Any) -> tuple[float, float] | None:
    if not isinstance(value, dict):
        return None
    try:
        point = float(value["x"]) / 1000, float(value["y"]) / 1000
    except (KeyError, TypeError, ValueError):
        return None
    if not all(math.isfinite(item) and 0 <= item <= 1 for item in point):
        return None
    return point


def _pressure_contains(pointer: tuple[float, float], t_ms: float, motion: dict[str, Any]) -> bool:
    phase = float(motion["phase_milliradians"]) / 1000
    angle = t_ms / float(motion["period_ms"]) * math.tau + phase
    center_x = .5 + float(motion["x_amplitude_milli"]) / 1000 * math.sin(angle)
    center_y = .5 + float(motion["y_amplitude_milli"]) / 1000 * math.sin(angle * 2 + phase * .63)
    dx = (pointer[0] - center_x) / (float(motion["hit_x_milli"]) / 1000)
    dy = (pointer[1] - center_y) / (float(motion["hit_y_milli"]) / 1000)
    return dx * dx + dy * dy <= 1 + 1e-6


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID:
        return _failure("mechanic mismatch")
    if str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return _failure("hidden mechanic mismatch")
    if str(public_state.get("mechanic_id") or "") != MECHANIC_ID:
        return _failure("public mechanic mismatch")
    if str(payload.get("challenge_id") or "") != str(ground_truth.get("challenge_id") or ""):
        return _failure("stale challenge")
    if str(public_state.get("challenge_id") or "") != str(ground_truth.get("challenge_id") or ""):
        return _failure("public and hidden challenge state disagree")
    task_id = str(ground_truth.get("task_id") or "")
    if not task_id or str(payload.get("task_id") or "") != task_id or str(public_state.get("task_id") or "") != task_id:
        return _failure("task identity mismatch")
    if public_state.get("board") != ground_truth.get("board"):
        return _failure("public and hidden course geometry disagree")
    motion = ground_truth.get("pressure_motion")
    if not isinstance(motion, dict) or public_state.get("pressure_motion") != motion:
        return _failure("public and hidden pressure motion disagree")

    board = ground_truth.get("board") or {}
    try:
        columns = int(board["columns"])
        rows = int(board["rows"])
    except (KeyError, TypeError, ValueError):
        return _failure("invalid hidden course")
    start = _point(board.get("start"))
    goal = _point(board.get("goal"))
    if start is None or goal is None:
        return _failure("invalid hidden endpoints")
    walls = {_point(item) for item in board.get("walls") or []}
    if None in walls:
        return _failure("invalid hidden wall")
    checkpoints = list(board.get("checkpoints") or [])
    checkpoint_positions = [_point(item) for item in checkpoints]
    checkpoint_ids = [str(item.get("id") or "") for item in checkpoints if isinstance(item, dict)]
    if any(item is None for item in checkpoint_positions) or len(checkpoint_ids) != len(checkpoints):
        return _failure("invalid hidden checkpoints")
    if checkpoint_ids != [str(item) for item in ground_truth.get("checkpoint_ids") or []]:
        return _failure("hidden checkpoint manifest disagrees with course geometry")

    events = payload.get("events")
    if not isinstance(events, list) or not events or len(events) > 800:
        return _failure("missing or oversized interaction transcript")

    position = start
    checkpoint_index = 0
    holding = False
    hold_start_t: float | None = None
    hold_start_seq: int | None = None
    resets = 0
    accepted_since_hold = 0
    last_pressure_sample_t: float | None = None
    pressure_samples = 0
    previous_t = -1.0
    move_directions: list[str] = []

    for index, event in enumerate(events):
        if not isinstance(event, dict):
            return _failure(f"event {index + 1} is not an object")
        try:
            seq = int(event.get("seq"))
        except (TypeError, ValueError):
            return _failure(f"event {index + 1} has no sequence")
        if seq != index + 1:
            return _failure("interaction sequence is not contiguous")
        event_t = _time(event.get("t_ms"))
        if event_t is None or event_t < previous_t:
            return _failure("interaction timestamps are invalid")
        previous_t = event_t
        event_type = str(event.get("type") or "")

        if event_type == "hold_start":
            if holding or position != start or checkpoint_index != 0:
                return _failure("pressure hold began from an invalid course state")
            holding = True
            hold_start_t = event_t
            hold_start_seq = seq
            accepted_since_hold = 0
            if _point(event.get("position")) != position:
                return _failure("hold-start position does not match replay")
            pointer = _normalized_pointer(event.get("pointer"))
            if pointer is None or not _pressure_contains(pointer, event_t, motion):
                return _failure("pressure hold did not begin on the moving plate")
            last_pressure_sample_t = event_t
            continue

        if event_type == "hold_sample":
            if not holding or last_pressure_sample_t is None:
                return _failure("pressure sample occurred outside an active hold")
            if event_t - last_pressure_sample_t > float(motion["maximum_sample_gap_ms"]):
                return _failure("moving pressure tracking has a sampling gap")
            pointer = _normalized_pointer(event.get("pointer"))
            if pointer is None or not _pressure_contains(pointer, event_t, motion):
                return _failure("pressure sample missed the moving plate")
            last_pressure_sample_t = event_t
            pressure_samples += 1
            continue

        if event_type == "hold_end":
            if not holding:
                return _failure("pressure release occurred without an active hold")
            if str(event.get("reason") or "") not in _RELEASE_REASONS:
                return _failure("unknown pressure release reason")
            if _point(event.get("position")) != position:
                return _failure("release position does not match replay")
            holding = False
            hold_start_t = None
            hold_start_seq = None
            position = start
            checkpoint_index = 0
            accepted_since_hold = 0
            last_pressure_sample_t = None
            resets += 1
            continue

        if event_type != "move":
            return _failure(f"unknown interaction event {event_type!r}")
        if not holding:
            return _failure("movement occurred while the pressure plate was released")
        if last_pressure_sample_t is None or event_t - last_pressure_sample_t > float(motion["maximum_sample_gap_ms"]):
            return _failure("vehicle moved without a recent moving-pressure sample")
        direction = str(event.get("direction") or "").upper()
        if direction not in _MOVES:
            return _failure("invalid movement direction")
        move_directions.append(direction)
        if _point(event.get("from")) != position:
            return _failure("reported movement origin does not match replay")
        dx, dy = _MOVES[direction]
        candidate = (position[0] + dx, position[1] + dy)
        accepted = (
            0 <= candidate[0] < columns
            and 0 <= candidate[1] < rows
            and candidate not in walls
        )
        if bool(event.get("accepted")) != accepted:
            return _failure("reported collision result does not match the course")
        if accepted:
            position = candidate
            accepted_since_hold += 1
            if checkpoint_index < len(checkpoint_positions) and position == checkpoint_positions[checkpoint_index]:
                checkpoint_index += 1
        if _point(event.get("to")) != position:
            return _failure("reported movement destination does not match replay")
        try:
            reported_checkpoint = int(event.get("checkpoint_index"))
        except (TypeError, ValueError):
            return _failure("movement is missing checkpoint progress")
        if reported_checkpoint != checkpoint_index:
            return _failure("reported checkpoint progress does not match replay")

    submitted_t = _time(payload.get("submitted_t_ms"))
    reported_duration = _time(payload.get("continuous_hold_duration_ms"))
    if submitted_t is None or reported_duration is None or submitted_t < previous_t:
        return _failure("completion timing is invalid")
    if not holding or hold_start_t is None or hold_start_seq is None:
        return _failure("the successful route was not covered by one continuous hold")
    replay_duration = submitted_t - hold_start_t
    if abs(reported_duration - replay_duration) > 250:
        return _failure("reported hold duration does not match the interaction transcript")
    if replay_duration < previous_t - hold_start_t:
        return _failure("pressure hold ended before the final move")
    if replay_duration < float(motion["minimum_hold_ms"]):
        return _failure("continuous moving-pressure interval was too short")
    if last_pressure_sample_t is None or submitted_t - last_pressure_sample_t > float(motion["maximum_sample_gap_ms"]):
        return _failure("moving pressure trace did not reach submission")
    if position != goal or checkpoint_index != len(checkpoint_positions):
        return _failure("course ended before every checkpoint and the dock")
    if accepted_since_hold < int(ground_truth.get("minimum_success_moves") or 0):
        return _failure("successful hold does not contain a complete traversable route")
    if payload.get("completed") is not True or payload.get("holding") is not True:
        return _failure("completion state was not active at submission")
    if _point(payload.get("final_position")) != position:
        return _failure("final position does not match replay")
    if _integer(payload.get("reset_count")) != resets:
        return _failure("reset count does not match releases")
    if _integer(payload.get("pressure_sample_count")) != pressure_samples:
        return _failure("pressure sample summary does not match replay")
    if _integer(payload.get("successful_hold_start_seq")) != hold_start_seq:
        return _failure("successful hold interval identity does not match replay")
    if [str(item) for item in payload.get("visited_checkpoints") or []] != checkpoint_ids:
        return _failure("checkpoint ledger does not match replay")
    if [str(item).upper() for item in payload.get("path") or []] != move_directions:
        return _failure("movement path does not match the full transcript")

    return {
        "graded": True,
        "passed": True,
        "score": 100,
        "feedback": (
            f"continuous hold covered {accepted_since_hold} accepted moves, "
            f"{len(checkpoint_ids)} checkpoints, and {resets} recoverable resets"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {
        "solution_path": list(ground_truth.get("solution_path") or []),
        "checkpoints": list((ground_truth.get("board") or {}).get("checkpoints") or []),
        "minimum_success_moves": int(ground_truth.get("minimum_success_moves") or 0),
    }
